### Biblioteker til oppgavene

<style>
    img{
        display: block;
        margin-left: auto;
        margin-right: auto;
    }
    img.flyt{
        display: inline;
        float: left
    }
    table {
        width:100%;
    }
</style>

In [2]:
import numpy as np
import scipy.special as sps

## 4.1.1 If-tester
### Oppgave 4.1
Lag et program som spør brukeren om alderen til brukeren. Skriv ut en kommentar som avhenger av om alderen er under eller over 17. Utvid programmet til å skille mellom flere aldre.

#### Løsning:

In [5]:
alder = int(input('Skriv inn alderen din: '))
grense = 17
                  
if alder < grense:
    print(f'Du er er under {grense} år.')
elif alder == grense:
    print(f'Du er er {grense} år gammel.')
else: 
    print(f'Du er over {grense} år.')

Du er er under 17 år.


### Oppgave 4.2
Lag et program som tar et tall som input. Programmet skal så vurdere og skrive ut om tallet er et positivt tall, negativt tall eller null.
#### Løsning:

In [6]:
tall = int(input('Skriv inn et heltall: '))

if tall < 0:
    print(f'Tallet {tall} er under 0.')
elif tall == 0:
    print(f'Tallet {tall} er lik 0.')
else: 
    print(f'Tallet {tall} er over 0.')

Tallet 42 er over 0.


### Oppgave 4.3
Lag et program som tar et tall som input. Programmet skal så vurdere og skrive ut om tallet er et heltall eller flyttall.
#### Løsning:

In [35]:
#Her er problemet at input tar kun string (tekst) som parameter. Metoden under fra boka (s56) vil ikke fungere.
tall = input('Skriv inn et tall: ')
if type(tall) != int:
    print(f'Tallet {tall} er ikke en int.')

Tallet 2 er ikke en int.


In [36]:
#Vi kan undersøke om tallet som brukeren tastet inn er et desimaltall
tall = float(input('Skriv inn et tall: '))
if tall == int(tall):
    tall = int(tall)  #Gjør det om til int hvis det er et heltall.
    print(f'Tallet {tall} er et heltall.')
else: 
    print(f'Tallet {tall} er et desimaltall.')

Tallet 2 er et heltall.


### Oppgave 4.4
Lag en funksjon som tar to argumenter som skal være tall. Funksjo- nen skal finne det minste tallet og skrive ut dette.
#### Løsning:

In [50]:
def minst_av_to(a, b):
    if a<b:
        return a
    elif a>b:
        return b
    
tall = minst_av_to(10, 3)
print(f'{tall} er det minste tallet.')

3 er det minste tallet.


### Oppgave 4.5
Lag en funksjon som tar to argumenter som skal være tall. Funk- sjonen skal avgjøre om det første tallet er minst og skrive ut True dersom dette stemmer.
#### Løsning:

In [54]:
def forst_er_storst(a, b):
    return a > b

tall = forst_er_storst(4, 2)
print(f'Det er {tall} at dt første tallet er størst.')

Det er True at dt første tallet er størst.


### Oppgave 4.6
Lag en funksjon som tar en poengsum som argument. Programmet skal finne ut hvilken karakter du får dersom maks poengsum er 60 poeng. Finn på karaktergrenser selv.
Utvid programmet til å ta maks poengsum som et tilleggsargument. Programmet
skal benytte prosenter av denne makssummen til å regne ut sluttkarakteren. Lag prosentgrensene selv.
Lag til slutt en funksjon som i tillegg tar alle poenggrensene som argumenter, enten som poeng eller prosenter av poengene.
#### Løsning:

In [71]:
def karakter(poeng):
    if poeng >= 57:
        return 6
    elif poeng >= 48:
        return 5
    elif poeng >= 37:
        return 4
    elif poeng >= 29:
        return 3
    elif poeng >= 18:
        return 2
    else:
        return 1

tildelt_karakter = karakter(48.5)
print(f'Karakterenr: {tildelt_karakter}.')

Karakterenr: 5.


### Oppgave 4.6 - Utvidet
#### Løsning:

In [7]:
def karakter(poeng, maks):
    faktor = poeng/maks
    if faktor >= 0.96:
        return 6
    elif faktor >= 0.80:
        return 5
    elif faktor >= 0.64:
        return 4
    elif faktor >= 0.48:
        return 3
    elif faktor >= 0.30:
        return 2
    else:
        return 1

tildelt_karakter = karakter(48.5, 100)  
print(f'Karakterenr: {tildelt_karakter}.')

Karakterenr: 3.


### Oppgave 4.7
Lag et program som spør om brukeren har Windows-PC eller Mac. Lag et oppfølgingsspørsmål på hver av disse, der du gir to eller tre alternativer. Du kan for eksempel spørre om hvorfor brukeren har Windows-PC eller Mac. Skriv ut en passende kommentar til svaret.
#### Løsning:

In [10]:
svar = input('Skriv "Windows-PC"hvis du har det og "Mac" hvis du har det. ')
if svar == 'Mac' or svar == 'Windows-PC':
    svar2 = input(f'Er du "fornøyd", "ikke-fornøyd", "vet-ikke" med {svar}-en: ')
else:
    print("Du hu har skrevet inn et ugyldig alternativ.")
    
if svar2 == 'fornøyd':
    print(f'Så bra du er fornøyd med {svar}-en!')
elif svar2 == 'ikke-fornøyd':
    print(f'Det var leit å høre at {svar}-en ikke oppfyller forventningene dine.')
elif svar2 == 'vet-ikke':
    print('Det må du finne ut av')
else:
    print("Du hu har skrevet inn et ugyldig alternativ.")

Så bra du er fornøyd med Mac-en!


### Oppgave 4.8
Gjør som i forrige oppgavene, men lag spørsmålene og alternativene helt selv. Prøv å få til tre nøstede if-setninger (if-setninger i if-setninger).
#### Løsning:

In [6]:
#Det er greit å øve på nestede if-setninger, men det er ikke ønskelig. 
#Koden blir vanskelig å lese og det finnes bedre alternativer

### Oppgave 4.9
Lag en kalkulator der brukeren kan taste inn to tall og velge regne- operasjon.
#### Løsning:

In [11]:
a = float(input('Skriv inn et tall: '))
b = float(input('Skriv inn et tall til: '))
operasjon = input('Hva vil du gjøre med tallene? (+ - * /): ')

if operasjon == '+':
    print(f'{a} {operasjon} {b} = {a+b}')
elif operasjon == '-':
    print(f'{a} {operasjon} {b} = {a-b}')
elif operasjon == '*':
    print(f'{a} {operasjon} {b} = {a*b}')
if operasjon == '/':
    print(f'{a} {operasjon} {b} = {a/b}')
else:
    print("Du har skrevet inn en ugyldig operasjon.")

4.0 * 42.0 = 168.0
Du har skrevet inn en ugyldig operasjon.


### Oppgave 4.10
Lag et lite eventyrspill der spilleren får ulike valg underveis, og historien formes av valgene spilleren tar. Det letteste er å lage en lineær historie, men du kan godt utfordre deg selv på å gjøre spillet ikke-lineært.
#### Løsning:

In [23]:
#;-)

### Oppgave 4.12
(Fysikk)

Lag et program der du kan taste inn bølgelengden til synlig lys i nm og få ut hvilken farge lyset har. Utvid eventuelt prorammet med andre deler av det elektromagnetiske spekteret.
<img src='http://docplayer.me/docs-images/17/80591/images/47-0.jpg' width='500px'>
#### Løsning:

In [13]:
bl = float(input("Skriv inn lysets bølgelengde: "))
if bl < 400 or bl >700:
    print(f'Lys med den oppgitte bølgelengden {bl} er ikke synlig')
elif bl >= 625:
    print(f'{bl} nm gir rødt lys.')
elif bl >= 590:
    print(f'{bl} nm gir oransje lys.')
elif bl >= 565:
    print(f'{bl} nm gir gult lys.')
elif bl >= 490:
    print(f'{bl} nm gir grønt lys.')
elif bl >= 440:
    print(f'{bl} nm gir blått lys.')
elif bl >= 420:
    print(f'{bl} nm gir indigo lys.')
elif bl >= 400:
    print(f'{bl} nm gir lilla lys.')


500.0 nm gir grønt lys.


### Oppgave 4.13 
Lag et program som spør brukeren om han/hun vil konvertere fra tommer (inch) til cm, eller fra cm til tommer. Dersom brukeren skriver noe annet, skal en feilmelding skrives ut.
Programmet skal så be brukeren om å skrive inn den ønskede størrelsen (i tommer eller cm) og deretter utføre den ønskede konverteringa. Du kan bruke at 1 tommer = 2.54 cm.
#### Løsning:

In [45]:
svar = input('Skriv "a" for å konvertere fra tommer til cm, og "b" for å konvertere fra cm til tommer. ')
if svar == 'a':
    tall = float(input('Skriv inn tallet du ønsker å konvertere. '))
    print(f'{tall}cm = {tall*2.54}tommer')
elif svar == 'b':
    tall = float(input('Skriv inn tallet du ønsker å konvertere. '))
    print(f'{tall}tommer = {tall/2.54:.2f}cm')
else:
    print(f'{svar} er en ugyldig kommando')


7.0tommer = 2.76cm


### Oppgave 4.14
(Matematikk R1)

Lag et program som regner ut funksjonen $g(x)=3x^4+5x^3+7x^2+x-3.0625$ or en verdi av x gitt av brukeren. Programmet skal skrive ut en melding om x − x-verdien fra brukeren er en faktor i g eller ikke.
#### Løsning:

In [58]:
x = float(input('Skriv inn x-verdi: '))
g = 3*x**4 + 5*x**3 + 7*x**2 + x - 3.0625
faktor = "er ikke"
if g == 0:
    faktor = "er"
    
print(f'g({x})={g}, (x-{x}) {faktor} en faktor i g.')

g(0.5)=0.0, (x-0.5) er en faktor i g.


### Oppgave 4.15
(Biologi): Hardy-Weinbergs-likevekt

<p> Vi skal se på en populasjon av mennesker og ser på et gen som finnes i to varianter: $a$ og $A$. Andelen av populasjonen som har variant a kan vi kalle for p og andelen av populasjonen som har variant $A$ kan vi kalle for $q$. Verdiene til $p$ og $q$ må være mellom $0$ og $1$. Siden alle i populasjonen vil ha enten $a$ eller $A$ (eller begge), må $p + q = 1$.</p>
<img src='https://upload.wikimedia.org/wikipedia/commons/b/b2/Hardy-Weinberg.svg' width='500px'>
<p>Populasjonen sies å være i Hardy-Weinberg-likevekt dersom verdiene til p og q ikke forandrer seg under forutsetningen at det foregår tilfeldig paring, ingen mutasjoner, ingen naturlig seleksjon og ingen evolusjon i populasjonen. Dersom populasjonen er i Hardy-Weinberg-likevekt kan en finne andelen til de tre mulige genotypene til menneskene i populasjonen:</p>
<ul>
    <li>$AA : p^2$</li>
    <li>$Aa : 2pq$</li>
    <li>$aa : q^2$</li>
</ul>
<p>Lag et program som ber brukeren om verdien til p og q. Programmet skal så teste om p + q = 1. Hvis summen er 1, skal programmet regne ut andelene til de tre mulige genotypene, og skrive dem ut. Hvis summen ikke er 1, skal programmet skrive en feilmelding.</p>
#### Løsning:

In [64]:
p = float(input('Skriv inn p-verdien: '))
q = float(input('Skriv inn q-verdien: '))
if p + q == 1:
    print(f'AA: {p**2:.2f}')
    print(f'Aa: {2*p*q:.2f}')
    print(f'aa: {q**2:.2f}')
else:
    print(f'{p} =! {q}')

AA: 0.09
Aa: 0.42
aa: 0.49


### Oppgave 4.16
(Fysikk): Heisenbergs uskarphetsrelasjon
 
Heisenbergs uskarphetsrelasjon forteller at det er ikke er mulig å ha en presis måling av en partikkels posisjon og bevegelsesmengde1samtidig. Det er altså en viss relasjon mellom hvor nøyaktig en måling kan være av posisjon og bevegelsesmengde. Matematisk kan en skrive dette som
$$\Delta x \cdot \Delta p \geq \frac{h}{2}$$
der vi bruker at $h = 1.055 · 10−34$ J, og $∆x$ og $∆p$ er mål på hvor nøyaktig en måling av henholdsvis posisjon og bevegelsesmengde er. Skriv et program som ber brukeren
om å skrive inn en verdi for ∆x og ∆p. Programmet skal så teste om Heisenbergs uskarphetsrelasjon vil holde for den gitte $∆x$ og $∆p$ og skrive ut en melding som forteller om relasjonen holder eller ikke. <b>Tips:</b> Et lite tall som f.eks $2.29 · 10^{−29}$ kan
representeres ved $2.29e-29$.
#### Løsning:

In [14]:
h = 1.055e-35
delta_x = float(input('Delta x: '))
delta_p = float(input('Delta p: '))

if delta_x*delta_p >= h/2:
    print('Heisenbergs uskarphetsrelasjon holder.')
else:
    print('Heisenbergs uskarphetsrelasjon holder ikke.')

Heisenbergs uskarphetsrelasjon holder.


### Oppgave 4.17
Kari liker å sove lenge. Hun blir veldig fornøyd hvis hun får sove 10 timer og ikke kommer for seint til skolen. For at hun ikke skal komme for seint, må hun seinest stå opp klokka 06:00.

#### a)
For at Kari skal sove til neste dag, må hun tidligst legge seg klokka 14:00. Lag  et program som ber brukeren om å skrive inn når Kari legger seg. Hun har litt OCD og legger seg kun ved hele timer, som er et heltall mellom 0 og 23. Programmet skal deretter teste om brukeren har gitt ei tid slik at Kari vil sove til neste dag. Hvis Kari ikke kommer til å sove til neste dag, skal programmet skrive en feilmelding.
#### Løsning:



In [88]:
leggetid = float(input('Når legger Kari seg? '))
vaknetid = (leggetid+10)%24
if leggetid >= 14:
    print(f'Kari vil sove til {vaknetid:02.0f}:00 neste dag')

Kari vil sove til 06:00 neste dag


#### b)
Utvid programmet ditt fra a) slik at det også tester om Kari vil rekke skolen
hvis hun sover 10 timer fra det angitte tidspunktet.
#### Løsning:

In [92]:
leggetid = float(input('Når legger Kari seg? '))
timer_sovn = 10
vaknetid = (leggetid + timer_sovn)%24
skole = 'rekke'

if leggetid >= 14:
    print(f'Kari vil sove til {vaknetid:02.0f}:00 neste dag med {timer_sovn} timer søvn.')
    
if vaknetid >= 6:
    skole = 'ikke rekke'
    
print(f'Kari vil {skole} skolen.')

Kari vil sove til 07:00 neste dag med 10 timer søvn.
Kari vil ikke rekke skolen.


### Oppfave 2.18
Skriv om kodesnuttene under slik det står beskrevet.

#### 1. 
Programmet skal ikke bruke and:

In [15]:
#Fra oppgaven
tall = 15
if 10 < tall and tall < 20:
    tall = tall + 2
else:
    tall = tall +1
print(tall)

17


In [16]:
#Løsning
tall = 15
if tall <= 10 or tall >=20:
    tall = tall +1
else:
    tall = tall +2
print(tall)

17


#### 2.  
På hver linje der det utføres +,-, * eller /, skal variabelen ikke forekomme på høyre side av likhetstegnet.

In [17]:
#Fra oppgaven
a = 10
print(a)
a = a + 14
print(a)
a = a/3 -2*a
print(a)

10
24
-40.0


In [18]:
#Løsning
a = 10
print(a)
a += 14
print(a)
a *= (1/3 - 2)
print(a)

10
24
-40.0


#### 3.
Variablene c (lysets hastighet), G (gravitasjonskonstanten) og Msol (solmassen i kg) skal skrives om til standardform:

In [19]:
#Fra oppgaven
c = 299792458
G = 0.0000000000667408
Msol = 1988550000000000000000000000000
print(c, G, Msol)

299792458 6.67408e-11 1988550000000000000000000000000


In [20]:
c = 299792458
G = 0.0000000000667408
Msol = 1988550000000000000000000000000
print(f' c    = {c:.2e} \n G    = {G:.2e} \n Msol = {Msol:.2e}' )

 c    = 3.00e+08 
 G    = 6.67e-11 
 Msol = 1.99e+30


### Oppgave 4.19
I denne oppgava skal du få prøve deg på å lage ei digital klokke som beregner hva klokka er etter ei gitt tid.
#### a)
Skriv et program som tar inn et klokkeslett fra brukeren på formen tt:mm der tt står for time (en verdi mellom 0 og 23) og mm står for minutt (en verdi mellom 0 og 59). Dette kan lettest gjøres ved at programmet spør om en verdi for tt, og så om mm. Programmet skal så sjekke om tt er en verdi mellom 0 og 23 og om mm er en verdi mellom 0 og 59. Hvis ikke, skal det skrive ut en feilmelding.
#### Løsning:

In [22]:
klokke = input('Skriv inn klokkeslettet (tt:mm)  ')
tt_ut = ''
mm_ut = ''

if klokke[0] != '0': 
    tt = int(klokke[0:2])
else:
    tt = int(klokke[1])

if klokke[3] != '0':
    mm = int(klokke[3:5])
else:
    mm = int(klokke[4])

if tt < 0 or tt > 23 or mm < 0 or mm > 59 or len(klokke) != 5:
    print('Du har skrevet inn noe feil. ')


#### b)
Utvidp rogrammet fra a) slik at dersom brukeren har gitt gyldig verdi for tt og mm, så ber den brukeren om å skrive inn hvor mange timer og minutter det har gått. Her skal det gå an å skrive inn hvilket som helst antall timer og minutter.

#### c)
Utvid programmet ditt slik at den finner hvilket klokkeslett den digitale klokka vil være på etter det gitte antallet timer og minutter. Det nye klokkeslettet skal også være på formen tt:mm. Det er et par ting programmet må ta høyde for:
-   Verdien til det nye tt må være mellom 0 og 23, og verdien til mm må være mellom 0 og 59. Her vil operatoren % være til veldig stor hjelp.
-   Dersom det nye antallet av minutter er mindre enn det gamle antallet minutter, betyr det at antall timer må økes med 1. Et eksempel er klokke- slettet 10:10; hvis det har gått 51 minutter, vil det nye klokkeslettet bli
11:01. Verdien for mm er nå 01, som er mindre enn 10.
-   Dersom antall minutter er over 60, vil det tilsvare én time. Programmet ditt må kunne finne ut hvor mange timer som utgjør det gitte antallet
minutter.

Programmet skal til slutt skrive ut det nye klokkeslettet.
#### Løsning:


In [23]:
klokke = input(' Skriv inn klokkeslettet (tt:mm) :')
tt_ut = ''
mm_ut = ''
tid_timer = 0
tid_minutter = 0
if klokke[0] != '0': 
    tt = int(klokke[0:2])
else:
    tt = int(klokke[1])

if klokke[3] != '0':
    mm = int(klokke[3:5])
else:
    mm = int(klokke[4])

if tt < 0 or tt > 23 or mm < 0 or mm > 59 or len(klokke) != 5:
    print('Du har skrevet inn noe feil. ') 
else:
    tid_timer = int(input(f' Hvor mange timer har det gått etter {tt:02.0f}:{mm:02.0f}  '))
    tid_minutter = int(input(f' Hvor mange minutter har det gått etter {tt:02.0f}:{mm:02.0f}  '))
    
tt_ny = tt + tid_timer 
mm_ny = mm + tid_minutter
tt_ny += mm//60 #Helstallsivisjon
tt_ny = tt_ny%24 #Resten av heltallsdivisjon
mm_ny = mm_ny%60
print(f'---- {tid_timer} timer og {tid_minutter} minutter etter {tt:02.0f}:{mm:02.0f} er klokken {tt_ny:02.0f}:{mm_ny:02.0f} ----')


---- 4 timer og 12 minutter etter 23:21 er klokken 03:33 ----


## 4.1.2 While-løkker
### Oppgave 4.20
Lag et motivasjonsprogram som skriver ut “Du er god!” et antall
ganger som er gitt av brukeren.
#### Løsning:

In [24]:
ant = int(input('Antall utskrifter: '))
i = 1
while i <= ant:
    print('Du er god! ')
    i+=1

Du er god! 
Du er god! 
Du er god! 
Du er god! 
Du er god! 


### Oppgave 4.21
Lag et program som skriver ut alle heltall fra og med 0 til og med 50. Bruk ei while-løkke.
#### Løsning:

In [25]:
n = 1
utskrift = ''
while n <= 50:
    utskrift = utskrift + str(n) + ' '
    n+=1
print(utskrift)

1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 


### Oppgave 4.22
Lag et program som skriver ut annethvert heltall fra 100 til 200 (ikke til og med). Bruk ei while-løkke.
#### Løsning:

In [26]:
n = 100
utskrift = ''
while n <= 200:
    utskrift = utskrift + str(n) + ' '
    n+=2
print(utskrift)

100 102 104 106 108 110 112 114 116 118 120 122 124 126 128 130 132 134 136 138 140 142 144 146 148 150 152 154 156 158 160 162 164 166 168 170 172 174 176 178 180 182 184 186 188 190 192 194 196 198 200 


### Oppgave 4.23 
Lag et program som skriver ut en tabell med overskriftene “tall” og “rota av tallet”. Skriv ut tallene fra og med 0 til og med 100 med de respektive røttene i kolonnen til høyre. Du trenger ikke skillelinjer i tabellen, men prøv å få til et pent
utskriftsformat.
#### Vertikal løsning:

In [27]:
linje = f'——————————————————————————————'
print(linje)
print(f'|{"Tall":^10}|{"Roten av tallet":^17}|')
n = 1
while n <= 10: #Bytt ut 10 med 100 for rett svar
    print( linje )
    print(f'|{n:^10.0f}|{n**0.5:^17.2f}|')
    n+=1
    
print(linje)

——————————————————————————————
|   Tall   | Roten av tallet |
——————————————————————————————
|    1     |      1.00       |
——————————————————————————————
|    2     |      1.41       |
——————————————————————————————
|    3     |      1.73       |
——————————————————————————————
|    4     |      2.00       |
——————————————————————————————
|    5     |      2.24       |
——————————————————————————————
|    6     |      2.45       |
——————————————————————————————
|    7     |      2.65       |
——————————————————————————————
|    8     |      2.83       |
——————————————————————————————
|    9     |      3.00       |
——————————————————————————————
|    10    |      3.16       |
——————————————————————————————


#### Horisontal løsning

In [28]:
linje = '———————————————————'
tall = f'| Tall             |'
roten_av_tall = f'| Roten av tallet  |'

n = 1
while n <= 10: 
    linje += '——————'
    tall += f'{n:4.1f} |'
    roten_av_tall += f'{n**0.5:4.1f} |'
    n+=1
    
print(linje)
print(tall)
print(linje)
print(roten_av_tall)
print(linje)

———————————————————————————————————————————————————————————————————————————————
| Tall             | 1.0 | 2.0 | 3.0 | 4.0 | 5.0 | 6.0 | 7.0 | 8.0 | 9.0 |10.0 |
———————————————————————————————————————————————————————————————————————————————
| Roten av tallet  | 1.0 | 1.4 | 1.7 | 2.0 | 2.2 | 2.4 | 2.6 | 2.8 | 3.0 | 3.2 |
———————————————————————————————————————————————————————————————————————————————


### Oppgave 4.24
Lag et program som skriver ut en tabell med overskriftene “tall” og “kvadratet av tallet”. Skriv ut annethvert tall fra og med 150 til (men ikke med) 200
med kvadratet i kolonnen til høyre.
#### Løsning:


In [29]:
linje = f'——————————————————————————————'
print(linje)
print(f'|{"Tall":^10}|{"Roten av tallet":^17}|')
n = 150
while n < 160: #Bytt 160 med 200 for rett svar
    print( linje )
    print(f'|{n:^10.0f}|{n**0.5:^17.2f}|')
    n+=1
    
print(linje)

——————————————————————————————
|   Tall   | Roten av tallet |
——————————————————————————————
|   150    |      12.25      |
——————————————————————————————
|   151    |      12.29      |
——————————————————————————————
|   152    |      12.33      |
——————————————————————————————
|   153    |      12.37      |
——————————————————————————————
|   154    |      12.41      |
——————————————————————————————
|   155    |      12.45      |
——————————————————————————————
|   156    |      12.49      |
——————————————————————————————
|   157    |      12.53      |
——————————————————————————————
|   158    |      12.57      |
——————————————————————————————
|   159    |      12.61      |
——————————————————————————————


### Oppgave 4.25
Lag en tabell som skriver ut antall grader fra og med 0 til og med 180, med 10 grader mellom hver verdi. I den andre kolonnen skal det være sinusverdien av vinkelen. Husk at standard vinkelmål i Python er radianer, så du må korrigere for dette under utregninga. Bruk tre gjeldende siffer.
#### Løsning:

In [30]:
linje = f'———————————————————————————————————————————'
print(linje)
print(f'|{"Vinkel i grader":^20}|{"Sinus til vinkelen":^20}|')
n = 0
while n <= 180: 
    print( linje )
    print(f'|{n:^20.0f}|{np.sin(2*np.pi*n/360):^20.2f}|')
    n+=10
    
print(linje)

———————————————————————————————————————————
|  Vinkel i grader   | Sinus til vinkelen |
———————————————————————————————————————————
|         0          |        0.00        |
———————————————————————————————————————————
|         10         |        0.17        |
———————————————————————————————————————————
|         20         |        0.34        |
———————————————————————————————————————————
|         30         |        0.50        |
———————————————————————————————————————————
|         40         |        0.64        |
———————————————————————————————————————————
|         50         |        0.77        |
———————————————————————————————————————————
|         60         |        0.87        |
———————————————————————————————————————————
|         70         |        0.94        |
———————————————————————————————————————————
|         80         |        0.98        |
———————————————————————————————————————————
|         90         |        1.00        |
————————————————————————————————

### Oppgave 4.26
Lag et program som regner ut summen av de 100 første naturlige tallene.

Et naturlig tall er i matematikken enten et positivt heltall (1, 2, 3,...) eller ikkenegativt heltall (0, 1, 2,...).

Den første definisjonen brukes oftest i tallteorien og skrives: $\mathbb{N}_0=\{0, 1, 2, 3, ...\}$

Den siste brukes innenfor predikatlogikk, mengdelære og datateknologi og skrives: $\mathbb{N}_1=\{1, 2, 3, ...\}$

#### Løsning:

In [31]:
#N_0
sum1 = 0
n = 0
while n < 100:
    sum1 += n
    n += 1
print(f'Summen av de hundre første tallene i N_0: {sum1}')

#N_1
sum2 = 0
n = 1
while n <= 100:
    sum2 += n
    n += 1
print(f'Summen av de hundre første tallene i N_1: {sum2}')

Summen av de hundre første tallene i N_0: 4950
Summen av de hundre første tallene i N_1: 5050


### Oppgave 4.27
Lag et program som regner ut denne summen:
$$\sum_{n=0}^{50} \frac{1}{2^n} $$
#### Løsning:

In [32]:
sum = 0
n = 0
while n <= 50:
    sum += 1/(2**n)
    n += 1
print(f'Summen har verdien {sum:.6f}')

Summen har verdien 2.000000


### Oppgave 4.28
Lag et program som regner ut denne summen:
$$\sum_{n=2}^{16} ((n^2+1)+n) $$
#### Løsning:

In [33]:
sum = 0
n = 2
while n <= 16:
    sum += ((n**2 + 1) + n)
    n+=1
print(f'Summen har verdien {sum:.1f}')

Summen har verdien 1645.0


### Oppgave 4.29
Lag et program som regner ut hvor lang tid det tar før du har en
viss sum i banken, regna ut fra en startinvestering og ei rente.
#### Løsning:

In [34]:
r = 3 #rente
vf = 1 + r/100 #vekstfaktor
bankbeholdning = 100000
sparemal = 200000
aar = 0
#Løsning med løkke
while bankbeholdning < sparemal:
    bankbeholdning *= vf
    aar += 1

print(f'Det vil ta {aar} år å oppnå en en bankbeholdning på {bankbeholdning:.2f} kr')

Det vil ta 24 år å oppnå en en bankbeholdning på 203279.41 kr


### Oppgave 4.30
Lag et program der brukeren skal logge seg inn med brukernavn og passord. Gi først brukeren uendelig mange sjanser til å skrive inn riktig brukernavn og passord. Når dette programmet virker, utvid programmet til å gi brukeren kun tre sjanser på å skrive inn riktig kombinasjon.
#### Løsning:

In [35]:
brukernavn = 'root'
passord = '123456789'
innlogging = True
antall_forsok = 1
while innlogging and antall_forsok <= 3:
    input_brukernavn = input('Tast inn ditt brukernavn: ')
    input_passord = input('Tast inn ditt passord: ')
    if input_brukernavn == brukernavn and input_passord == passord:
        innlogging = False
        print(f'Du er logget inn som {brukernavn} med passordet {passord}.')
    else:
        print('Brukernavn eller passord er feil. Prøv igjen.')
    antall_forsok += 1

Brukernavn eller passord er feil. Prøv igjen.
Du er logget inn som root med passordet 123456789.


### Oppgave 4.31

(utfordring)

Skriv et program som tar inn et heltall fra brukeren. Programmet skal så sjekke om heltallet inneholder sifferet 7 og skrive ut resultatet.

<b>Tips:</b> Her kan det være nyttig å bruke heltallsdivisjon // for å bevege seg gjen- nom tallets tierplass, hundreplass osv. Det vil også være nyttig å vite tallets største tierpotens, som kan finnes ved floor(log10(tall )). Funksjonen floor runder av et tall til nærmeste heltall som er mindre eller lik tallet selv.
#### Løsning:


In [38]:
antall_syvere = 0
input_heltall = int(input('Skriv inn et heltall:  '))
temp_heltall = input_heltall
n = np.floor(np.log10(input_heltall)) #Største tierpotens

while n >= 0:
    tall = temp_heltall // 10**n
    if tall == 7:
        antall_syvere += 1
    temp_heltall -= tall*10**n
    n -= 1
    
print(f'Tallet {input_heltall} inneholder {antall_syvere} 7-ere')

Tallet 98712382394782 inneholder 2 7-ere


#### Løsning med traversering av streng:

In [41]:
antall_syvere = 0
input_heltall = input('Skriv inn et heltall:  ')
lengde_til_heltall = len(input_heltall)
n = 0
while n < lengde_til_heltall:
    if input_heltall[n] == '7':
        antall_syvere += 1
    n += 1

print(f'Tallet {input_heltall} inneholder {antall_syvere} 7-ere')

Tallet 72368747 inneholder 3 7-ere


## 4.1.3 For-løkker
### Oppgave 4.32
Løs et utvalg av oppgavene 4.18–4.27 med for-løkker istedenfor
while-løkker.

In [408]:
# Prøv selv (i oppgavene 2.20 til 2.28 einer for-løkke seg bedre enn while-løkke)

### Oppgave 4.33
Lag et program som skriver ut:
1. Alle heltall fra og med 0 til og med 100.
2. Alle heltall fra 50 til 75.
3. Alle partall fra og med 0 til og med 100.
4. Alle oddetall fra og med 0 til og med 100.
5. (<b>Utfordring</b>) Alle primtall til og med 100 (hint: bruk heltallsdivisjon og/eller
modulus).

#### Løsning:

In [42]:
tall_1 = 'Heltall 0 til og med 100:\n' 
for x in range(101):
    tall_1 += str(x) + ' '
print(tall_1)

tall_2 = '\nHeltall 50 til og med 75:\n'
for x in range(50, 76):
    tall_2 += str(x) + ' '
print(tall_2)

tall_3 = '\nPartall 0 til og med 100:\n'
for x in range(0, 101, 2):
    tall_3 += str(x) + ' '
print(tall_3)

tall_4 = '\nOddetall 0 til og med 100:\n'
for x in range(1, 101, 2):
    tall_4 += str(x) + ' '
print(tall_4)

tall_5 = '\nPrimtall opp til 100:\n'
for x in range(2, 101):
    primtall = True
    for i in range (2,x+1):
        if x != i and x%i == 0:
            primtall = False
    if primtall:
        tall_5 += str(x) + ' '
print(tall_5)

Heltall 0 til og med 100:
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99 100 

Heltall 50 til og med 75:
50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 

Partall 0 til og med 100:
0 2 4 6 8 10 12 14 16 18 20 22 24 26 28 30 32 34 36 38 40 42 44 46 48 50 52 54 56 58 60 62 64 66 68 70 72 74 76 78 80 82 84 86 88 90 92 94 96 98 100 

Oddetall 0 til og med 100:
1 3 5 7 9 11 13 15 17 19 21 23 25 27 29 31 33 35 37 39 41 43 45 47 49 51 53 55 57 59 61 63 65 67 69 71 73 75 77 79 81 83 85 87 89 91 93 95 97 99 

Primtall opp til 100:
2 3 5 7 11 13 17 19 23 29 31 37 41 43 47 53 59 61 67 71 73 79 83 89 97 


### Oppgave 4.34
Hensikten bak programmet under er at det skal skrive ut hvert 3. tall fra og med 5 til 20. Men dersom vi kjører programmet, skrives ingenting ut! Forklar hvorfor og rett på programmet slik at det skriver ut hvert 3. tall fra og med 5 til 20.

In [ ]:
for tall in range(5,3,20):
    print(tall)

#### Løsning:

In [43]:
for tall in range(5,21,3):
    print(tall)

5
8
11
14
17
20


### Oppgave 4.35
Fibonnacifølgen er en kjent tallfølge med heltall der hvert tall etter
det første er summen av de to foregående. Følgen starter slik: 

1, 1, 2, 3, 5, 8, 13, 21, 34, 55, ...

Lag et program som finner tall nr. n i rekka.

#### Løsning:

In [44]:
n = 20
tall_1 = 1
tall_2 = 1
if n == 1 or n == 2:
    print(1)
else:
    for x in range(n-2):
        tall_n = tall_1 + tall_2
        tall_1 = tall_2
        tall_2 = tall_n
    print(f'Tall nr. {n} i fibonacifølgen er {tall_n}.')

Tall nr. 20 i fibonacifølgen er 6765.


### Oppgave 4.36
Lag et program som viser at summen av denne rekka er 4.5:

$$3 + 1 +\frac{1}{3} +\frac{1}{9}+\frac{1}{27}+...$$

#### Løsning:

In [45]:
sum = 0

for x in range(0, 10000):
    sum += 3/3**x
    
print(f'Summen til rekka er {sum:.10f}.')

Summen til rekka er 4.5000000000.


### Oppgave 4.37
Koden under regner ut:
$$\sum_{n=1}^m n = 1+2+\cdot \cdot \cdot + m$$

In [47]:
m = int(input(""))
s = 0
n = 1
while n <= m:
    s = s + n
    n = n + 1

Skriv om koden slik at den bruker en for-løkke, og slik at den skriver ut s.
#### Løsning:

In [48]:
m = int(input(" m: "))
s = 0
for n in range(1, m):
    s += n
print(f'Summen til rekka er {s} når m = {m}')

Summen til rekka er 3741 når m = 87


### Oppgave 4.38
I kombinatorikk er n-fakultet definert som produktet av alle heltall fra og med 1 til og med n. Dette kan vi skrive slik:
$$n!=\prod_{k=1}^n k$$
Lag en funksjon som regner ut fakultetet av et tall.
#### Løsning:

In [49]:
n = 11
n_fakultet = 1
for k in range(1, n+1):
    n_fakultet *= k
print(f'{n}! = {n_fakultet}')

11! = 39916800


### Oppgave 4.39
Lag et program som skal fungere som et ekko. Programmet skal be brukeren om å “si noe” (det vil si gi input), for så skrive ut det samme brukeren har sagt. Ekkoet skal slutte å virke når brukeren sier “slutt”.
#### Løsning:

In [51]:
input_tekst = ''
while True:
    input_tekst = input('Tekst: ')
    if input_tekst == 'slutt' or input_tekst == 'Slutt':
        break
    else:
        print(f'Ekko:  {input_tekst}')

Ekko:  Hva?
Ekko:  Hei


### Oppgave 4.40
Lag et program som finner alle tall som er delelig med 7, menikke med 5, mellom 2000 og 3200.
#### Løsning:

In [52]:
tall = ''
for x in range(2001, 3200):
    if x % 7 == 0 and x % 5 !=0:
        tall += str(x) + ' '
print(tall)

2002 2009 2016 2023 2037 2044 2051 2058 2072 2079 2086 2093 2107 2114 2121 2128 2142 2149 2156 2163 2177 2184 2191 2198 2212 2219 2226 2233 2247 2254 2261 2268 2282 2289 2296 2303 2317 2324 2331 2338 2352 2359 2366 2373 2387 2394 2401 2408 2422 2429 2436 2443 2457 2464 2471 2478 2492 2499 2506 2513 2527 2534 2541 2548 2562 2569 2576 2583 2597 2604 2611 2618 2632 2639 2646 2653 2667 2674 2681 2688 2702 2709 2716 2723 2737 2744 2751 2758 2772 2779 2786 2793 2807 2814 2821 2828 2842 2849 2856 2863 2877 2884 2891 2898 2912 2919 2926 2933 2947 2954 2961 2968 2982 2989 2996 3003 3017 3024 3031 3038 3052 3059 3066 3073 3087 3094 3101 3108 3122 3129 3136 3143 3157 3164 3171 3178 3192 3199 


### Oppgave 4.41
(Biologi)

Tøffeldyr formerer seg ukjønnet (deler seg i to) annenhver time dersom de befinner seg i en omgivelse med temperatur på $20°$C.

Skriv et program som regner ut og skriver ut hvor mange tøffeldyr det vil være dersom vi begynner med 5 tøffeldyr som formerer seg ukjønnet i løpet av 24 timer. Du kan anta at ingen tøffeldyr dør. 

<img src='http://biyolojiterimleri.com/f/images/K%20harfi/Kirpikliler.jpg' width='100px'>

#### Løsning:

In [53]:
# Oppgaven er litt upresis, men antar at tøffeldyrene akkurat har delt seg 
# og at de deler seg siste gang når det har gått 24 timer.
antall_toffeldyr = 5
for t in range(2, 25, 2): 
    antall_toffeldyr *= 2
print(f'Det vil være {antall_toffeldyr} tøffeldyr etter 24 timer')


Det vil være 20480 tøffeldyr etter 24 timer


### Oppgave 4.42
Dersom en PC-modell koster $7500$kr i dag, vil den nødvendigvis ikke koste det samme om noen år. Hvor mye en vare, som i vårt tilfelle er en PC- modell, avtar i verdi, bestemmes av ei rente r. Verdien P til PC-en etter t år kan finnes ved å bruke formelen
$$P=\frac{7500}{(1+r)^t}$$

Skriv et program som skriver ut, år for år, verdien til PC-en over en periode på $10$ år med ei rente på $4\%$.

In [54]:
r = 4/100
antall_aar = 10
pc_startverdi = 7500 
linje = f'—————————————————————————————————————————————————————'
print(linje)
print(f'|{"PC-en sin alder (år)":^25}|{"PC-en sin verdi (kr)":^25}|')
for t in range(0, antall_aar+1):
    print( linje )
    print(f'|{t:^25.0f}|{pc_startverdi/(1 + r)**t:^25.2f}|')
    n+=10
    
print(linje)

—————————————————————————————————————————————————————
|  PC-en sin alder (år)   |  PC-en sin verdi (kr)   |
—————————————————————————————————————————————————————
|            0            |         7500.00         |
—————————————————————————————————————————————————————
|            1            |         7211.54         |
—————————————————————————————————————————————————————
|            2            |         6934.17         |
—————————————————————————————————————————————————————
|            3            |         6667.47         |
—————————————————————————————————————————————————————
|            4            |         6411.03         |
—————————————————————————————————————————————————————
|            5            |         6164.45         |
—————————————————————————————————————————————————————
|            6            |         5927.36         |
—————————————————————————————————————————————————————
|            7            |         5699.38         |
————————————————————————————

### Oppgave 4.43

(Fysikk): Bohrs atommodell 

Bruk Bohrs atommodell for atomspektrene til hydrogen- atomet. Vi kan bruke formelen nedenfor for å regne ut frekvensen til et foton som emitteres når atomet deeksiterer fra skall n til m:

$$f=\frac{B}{h}\cdot \bigg( \frac{1}{m^2}-\frac{1}{n^2} \bigg)$$

Lag et program som regner ut alle bølgelengdene til fotonene som emitteres når atomet deeksiterer fra et skall n til alle mulige energinivåer.

<img class='flyt' src='http://ungenergi.no/wp-content/uploads/hydrogenatom.jpg' width='200px'>
<img class='flyt' src='https://upload.wikimedia.org/wikipedia/commons/1/17/Bohr_atom_animation_2.gif' width='200px'>
<img class='flyt' src='https://upload.wikimedia.org/wikipedia/commons/thumb/9/93/Bohr_atom_model.svg/620px-Bohr_atom_model.svg.png' width='200px'>


#### Løsning:

In [55]:
def f(m, n):
    B = 2.18e-18
    h = 6.64e-34
    return B/h*(1/m**2 - 1/n**2) 

def lamda(f):
    c = 3.0e8
    return 1e9*c/f
  
# n = int(input('Hva er skallet det skal deekseteres fra?'))
# m = int(input('Hva er skallet det skal deekseteres til?'))

print(f'Bølgelengden: {lamda(f(2, 4)):.3E}nm')

Bølgelengden: 4.873E+02nm


In [ ]:
def f(m, n):
    B = 2.18e-18
    h = 6.64e-34
    return B/h*(1/n**2 - 1/m**2) 

  
n = int(input('Hva er skallet det skal deekseteres fra?'))

print(linje)
print(f'|{"Fra skall":^15}:{n:^10}|')
linje = f'————————————————————————————'
print(linje)
print(f'|{"Skall":^12}|{"Energi (J)":^13}|')
for i in range(1, n):
    h = 6.64e-34
    if i != n:
        print( linje )
        print(f'|{i:^12.0f}|{h*f(n, i):12.3E} |')

print( linje )

### Oppgave 4.44
Over tid vil den radioaktive strålingen i et radioaktivt stoff reduseres. Dette skjer fordi atomene i det radioaktive stoffet vil omdannes til andre grunnstoffer. Mengden $N(t)$ som gjenstår av det radioaktive stoffet etter ei tid t kan finnes ved:

$$N(t)=N_0 e^{-at}$$

der $N_0$ er hvor mye radioaktivt stoff vi starter med ved tida $t = 0$. Verdien til a
forteller hvor raskt det radioaktive stoffet omdannes til andre grunnstoffer. Hvis vi
kjenner halveringstida $T$ , vil $a=\frac{ln(2)}{T}$

Plutonium-239 har halveringstid på omtrent $T = 24 000$ år. Skriv et program som skriver ut hvor mye Plutonium-239 som gjenstår etter hvert $5000$. år over en periode på $50000$ år dersom $N_0 = 4$kg.

#### Løsning:

In [519]:
T = 24000
a = np.log(2)/T
N_0 = 4
linje = f'———————————————————————'
print(linje)
print(f'|{"t (år)":^10}|{"N(t)":^10}|')
for t in range(0, 50001, 5000):
    print( linje )
    print(f'|{t:^10.0f}|{N_0*np.e**(-a*t):^10.2f}|')
    n+=10
    
print(linje) 

———————————————————————
|  t (år)  |   N(t)   |
———————————————————————
|    0     |   4.00   |
———————————————————————
|   5000   |   3.46   |
———————————————————————
|  10000   |   3.00   |
———————————————————————
|  15000   |   2.59   |
———————————————————————
|  20000   |   2.24   |
———————————————————————
|  25000   |   1.94   |
———————————————————————
|  30000   |   1.68   |
———————————————————————
|  35000   |   1.46   |
———————————————————————
|  40000   |   1.26   |
———————————————————————
|  45000   |   1.09   |
———————————————————————
|  50000   |   0.94   |
———————————————————————


### Oppgave 4.45
Per har $5000$kr på konto.For hvert år setter han inn $5\%$ avbeløpet på kontoen sin.

<b>a)</b> Skriv et program som finner ut hvor mye han vil ha på kontoen sin etter 20 år. Programmet skal skrive ut resultatet.

#### Løsning:

In [528]:
# a)
beholdning = 5000
antall_aar = 20
r = 5/100

for t in range(1, antall_aar+1):
    beholdning *= (1+r)

print(f'Per vil ha {beholdning:.2f}kr på kontoen etter {antall_aar}år')

Per vil ha 13266.49kr på kontoen etter 20år


<b>b)</b> Nå viser det seg at for hvert 5. år tar Per penger ut av kontoen. Han tar ut 10 % av beløpet på kontoen etter å ha lagt inn 5 % av beløpet som allerede ligger på kontoen. Utvid programmet ditt fra deloppgave a) slik det tar med uttakene Per gjør.

#### Løsning:

In [534]:
# b)
beholdning = 5000
antall_aar = 20
r = 5/100
uttak = 0

for t in range(1, antall_aar+1):
    beholdning *= (1+r)
    if t % 5 == 0:
        ut = beholdning * 10/100
        beholdning -= ut
        uttak += ut

print(f'Per vil ha {beholdning:.2f}kr på kontoen etter {antall_aar}år.'
      f'Han har tatt ut {uttak:.2f}kr' )

Per vil ha 8704.14kr på kontoen etter 20år.Han har tatt ut 3180.24kr


### Oppgave 4.46
(Biologi)

RNA er en kopi av en del av en DNA-streng (et gen). I translasjonen av et protein overfører tRNA aminosyrer til ribosomene. Hvert tRNA- molekyl har et antikodon som skal passe til mRNA-ets kodon.

<img width='100%' src='https://ndla.no/sites/default/files/images/proteinsyntesen.jpg?fag=7'>

1. Lag et program der du kan skrive inn tre baser på DNA som oversettes til mRNA-tråden.
2. Få programmet til å oversette kodonet til et antikodon.
3. Legg inn alle aminosyrene i programmet, og la programmet finne ut hvilken
aminosyre du får.
4. <b>NB: Omfattende!</b> Utvid programmet til å lese av lengre DNA-sekvenser.

#### Løsning:

In [623]:
input_DNA = input('DNA: ')
ut_mRNA = ''
for base in input_DNA:
    if base == 'A':
        ut_mRNA += 'U'
    if base == 'C':
        ut_mRNA += 'G'
    if base == 'G':
        ut_mRNA += 'C'
    if base == 'T':
        ut_mRNA += 'A'
        
print(f'DNA:   {input_DNA}')
print(f'm-RNA: {ut_mRNA}')

DNA:   TAACGCCATGTAAGCT
m-RNA: AUUGCGGUACAUUCGA


 <b>2.</b>  Få programmet til å oversette kodonet til et antikodon.
 #### Løsning:

In [627]:
input_DNA = input('DNA: ')
ut_mRNA = ''
for base in input_DNA:
    if base == 'A':
        ut_mRNA += 'U'
    if base == 'C':
        ut_mRNA += 'G'
    if base == 'G':
        ut_mRNA += 'C'
    if base == 'T':
        ut_mRNA += 'A'

ut_tRNA = ''
for base in ut_mRNA:
    if base == 'A':
        ut_tRNA += 'U'
    if base == 'C':
        ut_tRNA += 'G'
    if base == 'G':
        ut_tRNA += 'C'
    if base == 'U':
        ut_tRNA += 'A'
    
print(f'DNA:   {input_DNA}')
print(f'm-RNA: {ut_mRNA}')
print(f't-RNA: {ut_tRNA}')

DNA:   TAACGCCATGTAAGCT
m-RNA: AUUGCGGUACAUUCGA
t-RNA: UAACGCCAUGUAAGCU


<b>3.</b> Legg inn alle aminosyrene i programmet, og la programmet finne ut hvilken aminosyre du får. 
#### Løsning:

### Oppgave 4.47
Taylorrekke

I matematikken finnes det ulike måter å tilnærme en funksjon på. En metode som ofte brukes, kalles for rekkeutvikling (hvis du vil vite mer, kan du søke på Taylorrekke). Ved denne metoden uttrykkes funksjoner som en sum. Dersom en rekkeutvikler cosinus, vil en få at:

$$cos(x)=1+\sum_{n=1}^\infty \frac{(-1)^n}{(2n)!}x^{2n}$$

<img src='https://upload.wikimedia.org/wikipedia/commons/5/51/Taylor_cos.gif'>

Skriv et program som regner ut rekkeutviklingen til cosinus for en verdi av x der
$x \ne 0$. Datamaskinen kan ikke representere ∞, så du kan heller summere rekka opp til ei stor nok grense N. Kontroller resultatet ditt ved å regne ut $cos(x)$. Finn en verdi
for N som gir en god tilnærming til resultatet du fikk.

#### Løsning:

In [559]:
x = np.pi
N = 10
cos_x_taylor = 1
for n in range(1, N):
    cos_x_taylor += (-1)**n/(np.math.factorial(2*n))*x**(2*n)
print(f'cos(pi)_tayler = {cos_x_taylor:.7f} \ncos(pi)        = {np.cos(np.pi):.7f}')    

cos(pi)_tayler = -1.0000000 
cos(pi)        = -1.0000000


### Oppgave 4.48
(Fysikk)

Per har lagd en kulekanon og ønsker å prøve den ut. Kanonen er plassert 5 meter unna en vegg som har et hull. Hullet begynner 7.5 meter over bakken og slutter 8 meter over bakken. Per har regna seg fram til at når kula treffer veggen, vil kula være

$$-\frac{25}{2} \cdot \frac{g}{v^2\cos^2\theta}+5\tan \theta$$

meter over bakken. Kula skytes ut i en vinkel θ med en fart på v meter per sekund. Du kan bruke at g = 9.81m/s2. Per ønsker å treffe hullet, og derfor må han vite hvilken θ og v som må brukes.

<b>a)</b> Per har bestemt seg for at kanonen skal skyte ut kulene med en vinkel $60°$. Lag et program som tar inn skytefarten v som input fra brukeren og regner ut hvor høyt på veggen ballen treffer (bruk likningen over). Programmet skal skrive ut om ballen treffer hullet eller ikke.

#### Løsning:
Siden det ikke er oppgitt en radius på kula så antar jeg at den er veldig nærme 0 :)

In [579]:
theta = np.deg2rad(60) #Gjør om til radianer
g = 9.81
v = float(input('Skriv inn utgangshastighet på kula (m/s): '))
h = -25/2 * g/(v**2 * np.cos(theta**2))+5*np.tan(theta)

if h <= 0:
    print('Kula triller langs bakken inn til veggen.')
elif h <= 7.5:
    print(f'Kula treffer veggen under hullet, {h:.2f}m over bakken')
elif h >= 8:
    print(f'Kula treffer veggen over hullet eller går helt over veggen, {h:.2f}m over bakken')
else:
    print(f'Kula treffer i hullet {h:.2f}m over bakken.')

Kula treffer i hullet 7.99m over bakken.


<b>b)</b> Utvid programmet ditt fra a) slik at det tar inn skytevinkelenen i grader som input og finner ut om ballen vil treffe hullet eller ikke.

#### Løsning

In [577]:
g = 9.81
v = float(input('Skriv inn utgangshastighet på kula (m/s): '))
theta = np.deg2rad(float(input('Skriv inn vinkelen til kanonen (grader): '))) #Gjør om til radianer
h = -25/2 * g/(v**2 * np.cos(theta**2))+5*np.tan(theta)

if h <= 0:
    print('Kula triller langs bakken og treffer veggen.')
elif h <= 7.5:
    print(f'Kula treffer veggen under hullet, {h:.2f}m over bakken')
elif h >= 8:
    print(f'Kula treffer veggen over hullet eller går helt over veggen, {h:.2f}m over bakken')
else:
    print(f'Kula treffer i hullet {h:.2f}m over bakken.')

Kula treffer i hullet 7.60m over bakken.


### Oppgave 4.49
I kapittel 2 fikk du se en framgangsmåte på hvordan man kan konvertere et tall fra titallsystemet til totallsystemet (side ??). I programmet under er mange av stegene i framgangsmåten beskrevet som en algoritme, men ikke slik at programmet kan kjøres.

In [ ]:
def finn_stoerste_n(x):
    n = 0
    så lenge 2**n er mindre eller lik x
        n økes med en
    funksjon returner n-1
    
tall_ti = int(input(''))

sett n slik at n er den største potensen slik at 2**n <= tall_ti

tallet i starter på n, slutter på 0 og senkes ett steg om gangen:
    hvis 2**i er mindre eller lik tall_ti:
        tall_ti = differansen mellom gamle verdien av tall_ti og 2**i
        print(1)
    ellers:
        print(0)

Skriv om programmet slik at stegene som er beskrevet i koden, blir gjort om til kjørbar kode. Ingen innrykk i koden skal endres.

#### Løsning:

In [606]:
def finn_stoerste_n(x):
    n = 0
    while 2**n <= x:
        n+=1
    return n-1
    
tall_ti = int(input(' Gjør om fra titalls- til totalls-systemet: '))
tall_to_ut = ''
tall_ti_ut = tall_ti
n = int(np.floor(np.log2(tall_ti)))
for i in range(n, 0, -1):
    if 2**i <= tall_ti:
        tall_ti = tall_ti - 2**i
        tall_to_ut += '1'
    else:
        tall_to_ut += '0'

print(f'{tall_ti_ut}_ti = {tall_to_ut}_to')

2929292_ti = 101100101100101000110_to


### Oppgave 4.50
Newtons binomialformel

Newtons binomialformel forteller oss hvordan en kan regne ut $(a + b)^n$ for et heltall $n$ og vilkårlige tallverdier $a$ og $b$. Den sier at

$$(a+b)^n = \sum_{r=0}^n\binom{n}{r}a^{n-r}b^r$$

der $\binom{n}{r}=\frac{n!}{r!(n-r)!}$. For eksempel hvis $n=4$, så er

\begin{align}
(a+b)^4 & = \binom{4}{0}a^{4-0}b^0+\binom{4}{1}a^{4-1}b^1+\binom{4}{2}a^{4-2}b^2+\binom{4}{3}a^{4-3}b^3+\binom{4}{4}a^{4-4}b^4 \\
 & = \binom{4}{0}a^{4}b^0+\binom{4}{1}a^{3}b^1+\binom{4}{2}a^{2}b^2+\binom{4}{3}a^{1}b^3+\binom{4}{4}a^{0}b^4 \\
\end{align}

Lag et program som ber brukeren om verdier for $a$, $b$ og $n$. Programmet skal så regne ut $(a + b)^n$ og deretter skrive ut resultatet. Sjekk om resultatet stemmer ved å først legge sammen verdiene $a$ og $b$, for så opphøye resultatet i $n$.

<b>Tips:</b> For å regne ut 􏰔$n$􏰕, kan du bruke funksjonen $scipy.special.binom(n,r)$ fra $scipy.special-biblioteket$.


#### Løsning:

In [ ]:
#kommer

### Oppgave 4.51

Pascals trekant

Pascals trekant er en talltrekant der tallene i trekanten følger et bestemt mønster. På n-te rad er tallene som befinner seg inni (altså ikke langs kantene) trekanten lik summen av de to nærmeste tallene over seg.

<img src='https://ds055uzetaobb.cloudfront.net/image_optimizer/525270b426404c042a8a250de1eb8ffcf95c7d2e.png'>

Verdiene til den $n$-te raden i trekanten kan finnes ved:

$$\binom{n}{0} \hskip 0.5em \binom{n}{1} \hskip 0.5em \binom{n}{2} \hskip 0.5em \cdot\cdot\cdot \hskip 0.5em \binom{n}{n-2} \hskip 0.5em \binom{n}{n-1} \hskip 0.5em \binom{n}{n}$$

der $\binom{n}{r}=\frac{n!}{r!(n-r)!}$ for heltall r og n.

<b>a)</b>
Skriv et program som tar inn en verdi $n$ fra brukeren og skriver ut den $n$-te raden i Pascals trekant. Du kan bruke funksjonen scipy.special.binom(n,r) fra
scipy.special-biblioteket til å regne ut 􏰔$n$􏰕.

#### Løsning:

In [696]:
n = int(input('skriv inn raden du ønsker fra Pascaal´s trekant: '))
pascal_rad_ut = ''
for r in range(n+1):
    pascal_rad_ut += f'{sps.binom(n, n-r):^5n}'
print(f'{pascal_rad_ut:80}')

  1    5   10   10    5    1                                                    


<b>b)</b>
Utvid programmet ditt fra a) slik at det skriver ut trekanten fra og med 0-te rad til og med n-te rad. Et eksempel på output kan være:

<img src='https://upload.wikimedia.org/wikipedia/commons/c/ca/Pascal_triangle_small.png'>

#### Løsning:

In [694]:
for n in range(0, 12):
    pascal_rad_ut = ''
    for r in range(n+1):
        pascal_rad_ut += f'{sps.binom(n, n-r):^5n}'
    print(f'{pascal_rad_ut:^80}')

                                       1                                        
                                     1    1                                     
                                  1    2    1                                   
                                1    3    3    1                                
                             1    4    6    4    1                              
                           1    5   10   10    5    1                           
                        1    6   15   20   15    6    1                         
                      1    7   21   35   35   21    7    1                      
                   1    8   28   56   70   56   28    8    1                    
                 1    9   36   84   126  126  84   36    9    1                 
              1   10   45   120  210  252  210  120  45   10    1               
            1   11   55   165  330  462  462  330  165  55   11    1            
